api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+
+Original issue: [DynamoRIO/dynamorio#2985](https://github.com/DynamoRIO/dynamorio/issues/2985).
+
+# Design
+
+This required the addition of new support in various DynamoRIO components, like drreg, drx, drmgr and core DR. Multiple contributors worked on designing and implementing the required changes.
+
+
+## Scatter/gather Instruction Expansion
+
+Owner: [Hendrik Greving](https://github.com/hgreving2304)
+
+As described above, we can simplify work for DR clients by replacing the scatter and gather instructions with a functionally equivalent sequence of scalar stores and loads. The expanded sequence is the unrolled version of the following loop:
+```
+num_accesses = vector_size / element_size
+for i = 0, 1, ..., (num_accesses-1), do
+  extract mask for the ith access from mask reg or mask vector
+  if mask is set, then
+    extract ith element of index vector
+    compute address = base + ith index element
+    if instr_is_gather, then
+      load data from address into a scalar reg
+      insert scalar data into destination vector
+    else // instr_is_scatter
+      Extract scalar data from source vector to scalar reg
+      Store data from scalar reg to address
+    done
+    clear ith mask in mask reg or mask vector
+  done
+done
+```
+
+Due to the x86 ISA, the extraction/insertion of the scalar value from/to the vector may involve multiple steps, e.g. to extract a 32-bit scalar value from a 512-bit `zmm` reg, we first need to extract a 128-bit `xmm` from it.
+
+drmgr in DR provides multiple phases of instrumentation. This expansion is done in the first phase known as app2app. As the name suggests, this phase is intended to transform app instructions to equivalent instructions. For simplicity, we also separate out the scatter or gather instructions from its basic block and create a separate fragment with only the expanded sequence. The logic for expanding scatter and gather instructions is implemented in the drx extension library as [`drx_expand_scatter_gather`](https://github.com/DynamoRIO/dynamorio/blob/eb5d5af8e3444912c9f3f70e5ebf7969252ee4d6/ext/drx/drx.h#L538), and can be used by any client that needs it, including drcachesim. This support was added by [commit](https://github.com/DynamoRIO/dynamorio/commit/4359ef134e47942004c09db04b54593579763186).
+
+As an example, the following are the expansions of some instructions.
+
+Expansion for `vpgatherdd`: https://gist.github.com/abhinav92003/637c07bac785428550f05c18319f80c2
```

Comment:
Seems better to inline this here and be self-contained than rely on external content when it's just text

Reply:
Done